### PR TITLE
Replace "DataSet" with "Dataset" in Markdown files

### PR DESCRIPTION
**Pull request by deepyaman**
> **NOTE:** Kedro datasets are moving from `kedro.extras.datasets` to a separate `kedro-datasets` package in
> [`kedro-plugins` repository](https://github.com/kedro-org/kedro-plugins). Any changes to the dataset implementations
> should be done by opening a pull request in that repository.
## Development notes

~I'll update in source files separately,  if that works.~

Note that any changes in `kedro-datasets` are entirely separate.

## Checklist

- [x] Read the [contributing](https://github.com/kedro-org/kedro/blob/main/CONTRIBUTING.md) guidelines
- [x] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [x] Updated the documentation to reflect the code changes
- [x] Added a description of this change in the [`RELEASE.md`](https://github.com/kedro-org/kedro/blob/main/RELEASE.md) file
- [ ] Added tests to cover my changes

### CODE REVIEW

**Comment by astrojuanlu** [comment]
Thanks for this push @stichbury. Indeed, I think we underestimated how difficult would all this be and particularly on the code side we've been blocked because of CI issues and unexpected things.

Let's try to be a bit strategic (maybe we can sync this afternoon or later this week) on how do we want to proceed.

**Comment by deepyaman** [comment]
> Is the renaming done in the dataset package code already? (You mention doing the source separately so is that a no?).

I will do the stuff for source in `kedro` core repo shortly. 🤞 I've broken it up, also because I'm traveling, and larger PRs are harder.

Re dataset package, no, it will be done next (maybe this weekend? next week?). Since they're in separate repos, they do have to be done one by one, and I guess (as @astrojuanlu pointed out) it's been more challenging that anticipated.

**Comment by deepyaman** [comment]
@stichbury @astrojuanlu Actually, I was just wrapping up the changes in .py files, and it turns out we don't really have references to these in our docstrings? And I'll intentionally not change it in the tests to the non-deprecated version; can leave that as a test of deprecation, and update the tests, etc. once we've removed the old aliases for 0.19.

```diff
diff --git a/docs/source/conf.py b/docs/source/conf.py
index 5624e36e..22702666 100644
--- a/docs/source/conf.py
+++ b/docs/source/conf.py
@@ -348,16 +348,16 @@ def autolink_replacements(what: str) -> list[tuple[str, str, str]]:
     is a reStructuredText link to their documentation.
 
     For example, if the docstring reads:
-        This LambdaDataSet loads and saves ...
+        This LambdaDataset loads and saves ...
 
-    Then the word ``LambdaDataSet``, will be replaced by
-    :class:`~kedro.io.LambdaDataSet`
+    Then the word ``LambdaDataset``, will be replaced by
+    :class:`~kedro.io.LambdaDataset`
 
     Works for plural as well, e.g:
-        These ``LambdaDataSet``s load and save
+        These ``LambdaDataset``s load and save
 
     Will convert to:
-        These :class:`kedro.io.LambdaDataSet` load and save
+        These :class:`kedro.io.LambdaDataset` load and save
 
     Args:
         what: The objects to create replacement tuples for. Possible values
```

In the end, this was the only change I was going to make, but it doesn't seem to be important/not user-facing.

**Review by astrojuanlu — CHANGES_REQUESTED** [review]
To @stichbury's points:

> Sorry if I missed a meeting where this was discussed but while I'm supportive of this change, I don't understand/have missed how it's all tied together so we update everything at once.

I agree our efforts in this are have been kind of scattered. We're now tracking a more comprehensive plan in https://github.com/kedro-org/kedro/issues/2740 

> Also, the code in the starters will need updating too, correct?
> Plus the graphic in the Data Catalog section on the website if we update the Spark Dataset

Right. I just added those two to the plan.

> Is the renaming done in the dataset package code already?

I don't think this has been started yet.

> But I am still confused because if this drops in a release this week, what does that mean for users copying code snippets.
> And blog posts. And also, what would be the strategy for third party datasets, and those coming from teams like the SparkStreaming dataset and the Databricks managed table one.

I think these are very good points.

- If users find the new names but paste them in an old Kedro version, they'll get an unhelpful error. This is likely to happen, because many users will go to `/en/stable/`, rather than the appropriate version of the docs corresponding.
- Users might decide to use the new names, but they can't do it everywhere because the migration plan is incomplete, and inconsistencies remain.

The point that makes this more difficult is that part of these changes got already in `main` before we got a proper plan, which means that, as of now, old names will start emitting a `DeprecationWarning` in the latest version of Kedro. Reverting them is not impossible, but it can create some churn.

And on top of that, I contend that we should review our approach to these warnings, because as it stands now they can't even be silenced and it creates lots of noise on the Kedro CLI https://github.com/kedro-org/kedro/issues/2744

My proposal is:

- In the interest of avoiding having users that paste broken example code, let's hold off this pull request until we have clarity on what's needed for a proper migration ahead of 0.19, hence tackling gh-2740
- In the interest of minimizing the warning noise (that creates user confusion but also CI madness), for the upcoming release (that most realistically would be shipped tomorrow) _while_ not delaying the release of 0.18.11 any further, let's replace the `DeprecationWarning` that we're emitting with new names with a `PendingDeprecationWarning`, which is hopefully a small diff that buys us some time.

What do you think @deepyaman @noklam @stichbury ?

**Comment by deepyaman** [comment]
> * In the interest of minimizing the warning noise (that creates user confusion but also CI madness), for the upcoming release (that most realistically would be shipped tomorrow) _while_ not delaying the release of 0.18.11 any further, let's replace the `DeprecationWarning` that we're emitting with new names with a `PendingDeprecationWarning`, which is hopefully a small diff that buys us some time.

How does this help? Is it only because of what you mention in #2744? In that case, we can just remove the offending code instead of changing the warning type.

**Comment by astrojuanlu** [comment]
> How does this help? Is it only because of what you mention in https://github.com/kedro-org/kedro/issues/2744?

> > * In the interest of minimizing the warning noise (that creates user confusion but also CI madness), for the upcoming release (that most realistically would be shipped tomorrow) _while_ not delaying the release of 0.18.11 any further, let's replace the `DeprecationWarning` that we're emitting with new names with a `PendingDeprecationWarning`, which is hopefully a small diff that buys us some time.
> 
> How does this help? Is it only because of what you mention in #2744? In that case, we can just remove the offending code (a holdover from the last time a bunch of datasets were deprecated for new names) instead of changing the warning type.

Yes, `PendingDeprecationWarning`s are hidden by default, and will basically be invisible. Addressing gh-2744 could be a more sound alternative and it wouldn't need that much code, it just requires some discussion.

**Comment by deepyaman** [comment]
> > > * In the interest of minimizing the warning noise (that creates user confusion but also CI madness), for the upcoming release (that most realistically would be shipped tomorrow) _while_ not delaying the release of 0.18.11 any further, let's replace the `DeprecationWarning` that we're emitting with new names with a `PendingDeprecationWarning`, which is hopefully a small diff that buys us some time.
> > 
> > 
> > How does this help? Is it only because of what you mention in #2744? In that case, we can just remove the offending code (a holdover from the last time a bunch of datasets were deprecated for new names) instead of changing the warning type.
> 
> Yes, `PendingDeprecationWarning`s are hidden by default, and will basically be invisible. Addressing [gh-2744](https://github.com/kedro-org/kedro/issues/2744) could be a more sound alternative and it wouldn't need that much code, it just requires some discussion.

Yeah, I think that's a more sound alternative, and I didn't realize it required more discussion (already had made a PR before I saw this 😅). I believe it's just a holdover from changes made ages ago.

**Comment by stichbury** [comment]
Just to confirm that I'm A-OK with these suggestions. Thank you for the summary @astrojuanlu and thanks for your patience @deepyaman

File: docs/source/extend_kedro/custom_datasets.md
Comment (noklam) [inline docs/source/extend_kedro/custom_datasets.md:271]:
```suggestion
To use `PartitionedDataset` with `ImageDataset` to load all Pokemon PNG images, add this to the data catalog YAML so that `PartitionedDataset` loads all PNG files from the data directory using `ImageDataset`:
```

File: docs/source/tutorial/spaceflights_tutorial_faqs.md
Comment (noklam) [inline docs/source/tutorial/spaceflights_tutorial_faqs.md:28]:
```suggestion
DatasetNotFoundError: Dataset 'companies' not found in the catalog
```

File: docs/source/tutorial/spaceflights_tutorial_faqs.md
Comment (noklam) [inline docs/source/tutorial/spaceflights_tutorial_faqs.md:37]:
```suggestion
#### DatasetError: An exception occurred when parsing config for Dataset
```

File: docs/source/tutorial/spaceflights_tutorial_faqs.md
Comment (noklam) [inline docs/source/tutorial/spaceflights_tutorial_faqs.md:44]:
```suggestion
Object 'ParquetDataset' cannot be loaded from 'kedro_datasets.pandas'. Please see the
```

File: docs/source/tutorial/spaceflights_tutorial_faqs.md
Comment (noklam) [inline docs/source/tutorial/spaceflights_tutorial_faqs.md:42]:
```suggestion
DatasetError: An exception occurred when parsing config for Dataset
```

**Review by noklam — COMMENTED** [review]
Find a few `Dataset` reference left, maybe nee a global search for `DataSet` not just for the error?

**Comment by deepyaman** [comment]
> Find a few `Dataset` reference left, maybe nee a global search for `DataSet` not just for the error?

@noklam This is what I did:

image

but I guess need to find instances of "DataSet" (probably make those into "dataset", since they're not class references).

The reason I didn't do "*DataSet" -> "*Dataset" is because stuff like ParquetDataset (i.e. changes in Kedro-Datasets) are yet to be done.

**Review by noklam — APPROVED** [review]
Make sense, this will be short live anyway as we will update kedro dataset too, I am not concerned.

**Comment by astrojuanlu** [comment]
> The reason I didn't do "*DataSet" -> "*Dataset" is because stuff like ParquetDataset (i.e. changes in Kedro-Datasets) are yet to be done.

Should we prioritize https://github.com/kedro-org/kedro-plugins/pull/255 then?

File: docs/source/tutorial/spaceflights_tutorial_faqs.md
Comment (merelcht) [inline docs/source/tutorial/spaceflights_tutorial_faqs.md:23]:
```suggestion
#### DatasetNotFoundError: Dataset not found in the catalog
```

**Comment by stichbury** [comment]
I'm happy to approve these changes as the RTD build is succeeding. There is one conflict that needs resolving though @deepyaman

**Review by merelcht — COMMENTED** [review]
Happy to approve this when all suggestions and conflicts are addressed 👍

**Comment by astrojuanlu** [comment]
Happy to land this when all the comments are addressed 👍🏽

File: docs/source/deployment/argo.md
Comment (deepyaman) [inline docs/source/deployment/argo.md:27]:
```suggestion
- [All node input/output datasets must be configured in `catalog.yml`](../data/data_catalog.md#use-the-data-catalog-with-the-yaml-api) and refer to an external location (e.g. AWS S3); you cannot use the `MemoryDataset` in your workflow
```

File: docs/source/deployment/argo.md
Comment (deepyaman) [inline docs/source/deployment/argo.md:177]:
```suggestion
For the purpose of this walk-through, we will use an AWS S3 bucket for datasets; therefore `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY` environment variables must be set to have an ability to communicate with S3. The `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY` values should be stored in [Kubernetes Secrets](https://kubernetes.io/docs/concepts/configuration/secret/) (an example [Kubernetes Secrets spec is given below](#submit-argo-workflows-spec-to-kubernetes)).
```

File: docs/source/deployment/aws_batch.md
Comment (deepyaman) [inline docs/source/deployment/aws_batch.md:21]:
```suggestion
- [All node input/output datasets must be configured in `catalog.yml`](../data/data_catalog.md#use-the-data-catalog-with-the-yaml-api) and refer to an external location (e.g. AWS S3). A clean way to do this is to create a new configuration environment `conf/aws_batch` containing a `catalog.yml` file with the appropriate configuration, as illustrated below.
```

File: docs/source/extend_kedro/common_use_cases.md
Comment (deepyaman) [inline docs/source/extend_kedro/common_use_cases.md:7]:
```suggestion
The execution timeline of a Kedro pipeline can be thought of as a sequence of actions performed by various Kedro library components, such as the [datasets](/kedro_datasets), [DataCatalog](/kedro.io.DataCatalog), [Pipeline](/kedro.pipeline.Pipeline), [Node](/kedro.pipeline.node.Node) and [KedroContext](/kedro.framework.context.KedroContext).
```

File: docs/source/extend_kedro/common_use_cases.md
Comment (deepyaman) [inline docs/source/extend_kedro/common_use_cases.md:15]:
```suggestion
You can use [datasets](/kedro_datasets) to interface with various different data sources. If the data source you plan to use is not supported out of the box by Kedro, you can [create a custom dataset](custom_datasets.md).
```

File: docs/source/resources/glossary.md
Comment (deepyaman) [inline docs/source/resources/glossary.md:5]:
```suggestion
 The Data Catalog is Kedro's registry of all data sources available for use in the data pipeline. It manages loading and saving of data. The Data Catalog maps the names of node inputs and outputs as keys in a Kedro dataset, which can be specialised for different types of data storage.
```

File: docs/source/tutorial/spaceflights_tutorial_faqs.md
Comment (deepyaman) [reply docs/source/tutorial/spaceflights_tutorial_faqs.md:44]: `ParquetDataSet` isn't renamed yet.

**Comment by astrojuanlu** [comment]
RTD failure is unrelated

**Comment by deepyaman** [comment]
> RTD failure is unrelated

@astrojuanlu Nice, I was wondering about that. Can this just be merged then?

**Comment by astrojuanlu** [comment]
I don't have permissions for it, but in any case I think the idea is to address the problem separately, see https://github.com/kedro-org/kedro-plugins/issues/306

**Comment by astrojuanlu** [comment]
https://github.com/kedro-org/kedro/pull/2888 is going first, this will most likely need a rebase

**Comment by deepyaman** [comment]
```bash
find . -name '*.md' -print0 | xargs -0 sed -i "" "s/\([^A-Za-z]\)DataSet/\1Dataset/g"
find . -name '*.md' -print0 | xargs -0 sed -i "" "s/DataSetError/DatasetError/g"
find . -name '*.md' -print0 | xargs -0 sed -i "" "s/DataSetAlreadyExistsError/DatasetAlreadyExistsError/g"
find . -name '*.md' -print0 | xargs -0 sed -i "" "s/DataSetNotFoundError/DatasetNotFoundError/g"
find . -name '*.md' -print0 | xargs -0 sed -i "" "s/IncrementalDataSet/IncrementalDataset/g"
find . -name '*.md' -print0 | xargs -0 sed -i "" "s/PartitionedDataSet/PartitionedDataset/g"
find . -name '*.md' -print0 | xargs -0 sed -i "" "s/MemoryDataSet/MemoryDataset/g"
find . -name '*.md' -print0 | xargs -0 sed -i "" "s/LambdaDataSet/LambdaDataset/g"
find . -name '*.md' -print0 | xargs -0 sed -i "" "s/CachedDataSet/CachedDataset/g"
find . -name '*.md' -print0 | xargs -0 sed -i "" "s/AbstractDataSet/AbstractDataset/g"
find . -name '*.md' -print0 | xargs -0 sed -i "" "s/AbstractVersionedDataSet/AbstractVersionedDataset/g"
git checkout -- RELEASE.md
```

results in

**Pull request merged**
